crates/notion-core/src/tool/mod.rs

```diff
@@ -131,6 +133,61 @@ impl ToolSpec {
             package => ToolSpec::Package(package.into(), version),
         })
     }
+
+    /// Get a valid, sorted `Vec<ToolSpec>` given a `Vec<String>`.
+    ///
+    /// Accounts for the following error conditions:
+    ///
+    /// - `notion install node 12`, where the user intended to install `node@12`
+    ///   but used syntax like in nodenv or nvm
+    /// - invalid version specs
+    ///
+    /// Returns a listed sorted so that if `node` is included in the list, it is
+    /// always first.
+    pub fn from_strings<T>(tool_strs: &mut [T], action: &str) -> Fallible<Vec<ToolSpec>>
+    where
+        T: AsRef<str>,
+    {
+        let mut iter = tool_strs.iter();
+        Self::check_args(iter.by_ref(), action)?;
+
+        let mut tools = iter
```

Comment:
I think you actually need to use a new iterator here. Even though you're only borrowing it in the call to `check_args`, it's being advanced by the calls to `.next()`, so when you get to this point it's missing 3 elements.

Reply:
Ahhhh, good catch, I misunderstood what I was reading in the docs. I'll just drop the last commit and then add the test for the positive case below. Thank you!